### PR TITLE
Fix ?(#TAB method search name exploration

### LINKED ISSUE #52551: Typing `?(` in REPL help mode causes exception
if I type `?(` when in the REPL help mode (i.e., after an initial `?`), I get the following error message

```
/tmp/julia-15244663a3/bin> ./julia
               _
   _       _ _(_)_     |  Documentation: https://docs.julialang.org
  (_)     | (_) (_)    |
   _ _   _| |_  __ _   |  Type "?" for help, "]?" for Pkg help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 1.11.0-DEV.1089 (2023-12-13)
 _/ |\__'_|_|_|\__'_|  |  Commit 15244663a34 (2 days old master)
|__/                   |

help?> ?(┌ Error: Error in the keymap
│   exception = (UndefVarError(Symbol("@assume_effects"), Main), Union{Ptr{Nothing}, Base.InterpreterIP}[Ptr{Nothing} @0x00007f0753c39c6e, Ptr{Nothing} @0x00007f0753bd1c58, Ptr{Nothing} @0x00007f06884d87a0, Ptr{Nothing} @0x00007f0688730465, Ptr{Nothing} @0x00007f06884fd041, Ptr{Nothing} @0x00007f0688642463, Ptr{Nothing} @0x00007f0753bb102d, Ptr{Nothing} @0x00007f0688550c38, Ptr{Nothing} @0x00007f06885cc25e, Ptr{Nothing} @0x00007f0753bb102d, Ptr{Nothing} @0x00007f0753bc1b99, Ptr{Nothing} @0x00007f068860a31c, Ptr{Nothing} @0x00007f0688520f36, Ptr{Nothing} @0x00007f0753bb102d, Ptr{Nothing} @0x00007f06886b8ed5, Ptr{Nothing} @0x00007f068853029d, Ptr{Nothing} @0x00007f06885d3f3e, Ptr{Nothing} @0x00007f0753bb102d, Ptr{Nothing} @0x00007f06884fcb49, Ptr{Nothing} @0x00007f06885f6acd, Ptr{Nothing} @0x00007f06885d5c06, Ptr{Nothing} @0x00007f0753bb102d, Ptr{Nothing} @0x00007f0753bd63e1])
└ @ REPL.LineEdit /tmp/julia-15244663a3/share/julia/stdlib/v1.11/REPL/src/LineEdit.jl:2821

```

### PR DESCRIPTION
**Pull request by Liozou**
Fix #52551.

This PR ensures that a `SomeModule.?(...#TAB` completion can only suggests method `foo` such that `SomeModule.foo` exists (by checking `isdefined(SomeModule, :foo)`). This is equivalent, I believe, to the initial implementation of https://github.com/JuliaLang/julia/pull/38791, less the bug.

Now that we have #51345, we may want to relax the above condition somewhat to include public names present in modules loaded into `SomeModule`, so that, for instance, a direct completion of `?(` would include `@assume_effects`. This could be good for method exploration because even though typing `@assume_effects` with no qualification in `Main` will error, the error now includes the helpful message
```
Hint: a global variable of this name also exists in Base.
```
But that can wait for a later PR anyway, this one is just the bugfix.

The bug mentioned at https://github.com/JuliaLang/julia/issues/52551#issuecomment-1858543413 dates from the initial #38791 so this could be backported as far back as v1.8.

### CODE REVIEW

File: stdlib/REPL/src/REPLCompletions.jl [hunk 1]
```diff
@@ -713,6 +713,21 @@ function complete_methods(ex_org::Expr, context_module::Module=Main, shift::Bool
...
+            if !isa(func, Module)
+                funct = Core.Typeof(func)
+                push!(seen, funct)
+            elseif isa(func, Module) && Symbol(func) ∉ exploredmodules
+                recursive_explore_names!(seen, exploredmodules, func, initial_module)
+            end
+        end
+    end
+end
+
```
Comment (aviatesk) [inline stdlib/REPL/src/REPLCompletions.jl:730]:
```suggestion
function recursive_explore_names!(seen::Base.IdSet{Any}, callee_module::Module, initial_module::Module, 
                                  exploredmodules::IdSet{Module}=IdSet{Module}())
    push!(exploredmodules, callee_module)
    for name in names(callee_module; all=true, imported=true)
        if !Base.isdeprecated(callee_module, name) && !startswith(string(name), '#') && isdefined(initial_module, name)
            func = getfield(callee_module, name)
            if !isa(func, Module)
                funct = Core.Typeof(func)
                push!(seen, funct)
            elseif isa(func, Module) && func ∉ exploredmodules
                recursive_explore_names!(seen, func, initial_module, exploredmodules)
            end
        end
    end
    return seen
end

function recursive_explore_names(callee_module::Module, initial_module::Module)
    seen = Base.IdSet{Any}()
    recursive_explore_names!(seen, callee_module, initial_module)
    return seen
end
```
and using `recursive_explore_names` within `complete_any_methods` would be more correct and cleaner?

Comment (Liozou) [reply stdlib/REPL/src/REPLCompletions.jl]: Absolutely, that's better! Thanks, done in https://github.com/JuliaLang/julia/pull/52555/commits/46193b842f73b518018b77020dc07d5c35c5fb5e

**Comment by aviatesk** [comment]
Thanks for fixing this!

**Review by aviatesk — APPROVED** [review]

**Comment by Liozou** [comment]
With pleasure! Should it be backported to 1.10?

**Comment by aviatesk** [comment]
Yeah, I think this is eligible for backport. Still, it might be wise to proceed in a manner that doesn't hold up the release, like aiming for 1.10.1 instead of 1.10.0.

**Comment by aviatesk** [comment]
Removing the backport label.

**Pull request merged**